### PR TITLE
Vol2 hpo

### PR DESCRIPTION
**Pull request by aaronkl**
*Description of changes:*

Add initial set of notebooks for the HPO chapter: index, intro, api, hyperband, async RS, async SH.
 

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mseeger** [comment]
Aaron is OOTO from Wed, I'd be the PoC then

**Comment by AnirudhDagar** [comment]
Thanks @mseeger, we'll get the PR reviewed today itself and hopefully merged soon as well.

File: chapter_hyperparameter_optimization/sh-async.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+trial_backend = PythonBackend(tune_function=objective, config_space=config_space)
+
+stop_criterion = StoppingCriterion(
+    max_wallclock_time=max_wallclock_time
+)
+tuner = Tuner(
+    trial_backend=trial_backend,
+    scheduler=scheduler,
+    stop_criterion=stop_criterion,
+    n_workers=n_workers,
```
Comment (AnirudhDagar) [inline chapter_hyperparameter_optimization/sh-async.md:151]:
```suggestion
    n_workers=n_workers
```

File: chapter_hyperparameter_optimization/sh-async.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+```{.python .input  n=56}
+scheduler = ASHA(
+    config_space,
+    metric="validation_error",
+    mode="min",
+    max_resource_attr="max_epochs",
+    resource_attr="epoch",
+    type="promotion",
+    grace_period=1,  # this corresponds to r_min 
+    reduction_factor=2,  # this corresponds to eta
```
Comment (AnirudhDagar) [inline chapter_hyperparameter_optimization/sh-async.md:130]:
```suggestion
    reduction_factor=2  # this corresponds to eta
```

File: chapter_hyperparameter_optimization/sh-async.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    from d2l import torch as d2l    
+    from syne_tune import Reporter
+    model = d2l.AlexNet(lr=learning_rate)
+    trainer = d2l.Trainer(max_epochs=1, num_gpus=1)
+    data = d2l.FashionMNIST(batch_size=batch_size, resize=(224, 224))
+    report = Reporter() 
+    for epoch in range(1, max_epochs + 1):
+        trainer.fit(model=model, data=data)
+        validation_error = d2l.numpy(trainer.validate().cpu())
+        report(epoch=epoch-1, validation_error=float(validation_error))
```
Comment (AnirudhDagar) [inline chapter_hyperparameter_optimization/sh-async.md]:
```suggestion
    for epoch in range(max_epochs):
        trainer.fit(model=model, data=data)
        validation_error = d2l.numpy(trainer.validate().cpu())
        report(epoch=epoch, validation_error=float(validation_error))
```
We can probably simplify this, right?

File: chapter_hyperparameter_optimization/lib.bib [hunk 4]
```diff
@@ -0,0 +1,1713 @@
+@inproceedings{klein-ejs17,
```
Comment (AnirudhDagar) [inline chapter_hyperparameter_optimization/lib.bib:1]: Let's move the contents of this file to `d2l.bib`.

**Review by AnirudhDagar — CHANGES_REQUESTED** [review]
@aaronkl Overall this PR looks extremely polished already; thanks for putting in the effort! :))

I've shared my comments below in the review, most of them are small. Some major comments are:

1. Let's reuse models from previous chapters instead of defining them again and saving them in HPO. We can save them in the previous chapter itself. Eg Alexnet, SoftmaxClassification

2. I can see that the CI is taking a lot of time to run the following notebooks:
     * `hyperopt-api.md`: >17min
     * `hyperopt-intro.md`: timing out i.e more than 20min
      * `hyperband-intro.md`: >15min

Ideally, we'd want to reduce this time but I can understand running randomsearch etc. will naturally take time, based on the number of trials. A few suggestions to improve the timings are to consider a smaller simpler model and maybe reduce the trials. Anything above 15min can cause timeouts later if we have simultaneous jobs running on the GPUs. For reference the longest notebook cnn-design takes about 10-12min right now. All other notebooks on an avg run no longer than 7-8 minutes.

3. Let's move bib file contents to d2l.bib instead of a separate file for this chapter only.

File: chapter_hyperparameter_optimization/rs-async.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+from syne_tune.backend.python_backend import PythonBackend
+from syne_tune.optimizer.baselines import RandomSearch
+from syne_tune import Tuner, StoppingCriterion
+from syne_tune.experiments import load_experiment
+from syne_tune.report import Reporter
+
+
+def objective(learning_rate, batch_size, max_epochs):
+    from d2l import torch as d2l    
+    from syne_tune import Reporter
```
[the remark this answers is not in the document]
Comment (aaronkl) [reply chapter_hyperparameter_optimization/rs-async.md:61]: Actually these imports need to be inside the function since Syne Tune serializes this function later

File: chapter_hyperparameter_optimization/hyperopt-api.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                tf.keras.layers.Dense(4096, activation='relu'),
+                tf.keras.layers.Dropout(0.5),
+                tf.keras.layers.Dense(num_classes)])
+```
+
+For that, we first have to define the objective function.
+
+```{.python .input  n=9}
+def objective(batch_size, learning_rate, max_epochs=8):  #@save
+    model = d2l.AlexNet(lr=learning_rate)
```
[the remark this answers is not in the document]
Comment (aaronkl) [reply chapter_hyperparameter_optimization/hyperopt-api.md:227]: Yes that's a good point. Any suggestions which model to take?

**Comment by astonzhang** [comment]
Thanks. As @AnirudhDagar pointed out, it would be ideal to reduce the running time of any notebook to < 10 mins. E.g., if this is obtained by reducing training epochs, we may encourage readers to increase it themselves in Exercises.

File: chapter_hyperparameter_optimization/hyperopt-api.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (AnirudhDagar) [reply chapter_hyperparameter_optimization/hyperopt-api.md:227]: I'd suggest using Lenet due to its simplicity and training time being much less.

**Review by astonzhang — COMMENTED** [review]
Thanks for the amazing work! Besides comments, here are my initial feedback:

* For this initial PR, add PyTorch implementation only
* It would be ideal to reduce the running time of any notebook to < 10 mins (Consider Simpler backbone such as LeNet).  E.g., if this is obtained by reducing training epochs, we may encourage readers to increase it themselves in Exercises.
* Subclass d2l.Trainer, say d2l.HPOTrainer where you add `validate` method or call it sth else (see comments).
* Figures should have transparent background
* Once all done, make sure to run `d2lbook build lib` to save all classes/functions with the comment #@save to the `d2l/*` lib where you can invoke them as `d2l.XXX` (pip install git+https://github.com/d2l-ai/d2l-book)

TODO:
* Please revert changes to static/frontpage/*
* Add missing exercises (at least 2 for each notebook)

I may add more feedback after the CI passes to build the website.

File: chapter_hyperparameter_optimization/hyperopt-intro.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+Faced with all these challenges, we usually try to identify a small set of well performing hyperparameter configurations quickly, instead of hitting the global optima exactly. However, due to large computational demands of most neural networks models, even this can take days or weeks of compute. We will explore in :numref:`sec_mf_hpo` how we can speed-up the optimization process by either distributing the search or using cheaper-to-evaluate approximations of the objective function.
+
+
+Now, since we would like to optimize the validation error, we need to add a function computing this quantity.
+
+```{.python .input  n=8}
+%%tab pytorch
+
+@d2l.add_to_class(d2l.Trainer)  #@save
+def validate(self):
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_hyperparameter_optimization/hyperopt-intro.md:91]: Instead of directly modifying the existing Trainer class in the d2l lib, please:
* subclass d2l.Trainer, say HPOTrainer. If you need to use it elsewhere, save it so you have d2l.HPOTrainer
* Implement `validate` in `d2l.HPOTrainer` or call it `validation_step` to be consistent in naming

[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_hyperparameter_optimization/hyperopt-intro.md:91]: run `d2lbook build lib`: HPOTrainer will be saved to d2l/torch.py

File: chapter_hyperparameter_optimization/hyperopt-api.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+                tf.keras.layers.Conv2D(filters=256, kernel_size=3, padding='same',
+                                       activation='relu'),
+                tf.keras.layers.MaxPool2D(pool_size=3, strides=2),
+                tf.keras.layers.Flatten(),
+                tf.keras.layers.Dense(4096, activation='relu'),
+                tf.keras.layers.Dropout(0.5),
+                tf.keras.layers.Dense(4096, activation='relu'),
+                tf.keras.layers.Dropout(0.5),
+                tf.keras.layers.Dense(num_classes)])
+```
```
[the remark this answers is not in the document]
Comment (mseeger) [reply chapter_hyperparameter_optimization/hyperopt-api.md:220]: OK, and I'll use `LeNet`

File: chapter_hyperparameter_optimization/hyperopt-api.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mseeger) [reply chapter_hyperparameter_optimization/hyperopt-api.md:227]: OK

File: chapter_hyperparameter_optimization/hyperopt-intro.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+```{.python .input  n=9}
+%%tab pytorch
+class SoftmaxClassification(d2l.Classifier): #@save
+    def __init__(self, num_outputs, lr):
+        super().__init__()
+        self.save_hyperparameters()
+        self.net = nn.Sequential(nn.Flatten(),
+                                 nn.LazyLinear(num_outputs))
+    def forward(self, X):
+        return self.net(X)
```
[the remark this answers is not in the document]
Comment (mseeger) [reply chapter_hyperparameter_optimization/hyperopt-intro.md:69]: OK

File: chapter_hyperparameter_optimization/rs-async.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mseeger) [reply chapter_hyperparameter_optimization/rs-async.md:67]: I'll add a comment

**Comment by AnirudhDagar** [comment]
Yes that is the reason why all the models in Modern Conv chapter are slower than LeNet where we don't scale the input images.

**Comment by mseeger** [comment]
OK, I think the examples are probably fast enough now. I also changed the code so that default config (lr=0.1, batch_size=128) which is used in the earlier chapter, is always the first one in all HPO algorithms.

It may be that the example is too simple to differentiate between the HPO algorithms. This is the usual danger with
running HPO for too little time. I'll see what can be done.

**Comment by astonzhang** [comment]
@mseeger Thanks. Can you fix the CI error: http://ci.d2l.ai/blue/organizations/jenkins/d2l-en/detail/PR-2372/36/pipeline

**Comment by astonzhang** [comment]
You may click "Details" beside the red cross to view the error message. Once the CI passes, I'll take a look. Thanks!

Screen Shot 2022-12-08 at 12 09 44 AM

**Comment by mseeger** [comment]
CI fails now not due to me, I think:

Convert _build/eval_all/chapter_linear-networks/linear-regression.ipynb to _build/rst_all/chapter_linear-networks/linear-regression.rst

**Comment by mseeger** [comment]
I just added one #@add tag in that section

**Comment by AnirudhDagar** [comment]
@mseeger I noticed that `hyperopt-intro` and `hyperband-intro` notebooks are running on CPU only. I don't think that is intended. Please correct me if I'm wrong. That is causing slowdowns.

This is happening because d2lbook [intelligently](https://github.com/d2l-ai/d2l-book/blob/a1ce88f3894cf9b5714189fd774b5fb9d9a7c6a6/d2lbook/resource.py#L30-L46) (not so intelligent though just some string matching) tries to determine if a notebook requires GPUs so that it can distribute CI resources more efficiently (no point in wasting GPU resources for a notebook that is just as fast on CPU eg. chapter_preliminaries), in these two notebooks it is unable to find any marks to assign GPU resources to evaluate the notebook. 

We have a TODO to improve that on our infra because as you can see it is not very accurate.

For now, just to see if things are significantly faster by forcing d2lbook to recognize notebook needs GPU, I can add a small hack in the notebook itself. But before I do that please let me know if these two notebooks need to run on GPU or CPU. I assume running on GPU will speed them up significantly.

You can [see](http://ci.d2l.ai/blue/organizations/jenkins/d2l-en/detail/PR-2372/44/pipeline/#step-12-log-204) they use only a CPU resource:
image

**Comment by mseeger** [comment]
@AnirudhDagar Yes I also noted that the notebooks run on CPU only. Cool if you could fix this. It may make HPO look better, because it can get more evaluations done.

I am now pretty happy with it, having fixed many things. I'll still look into some details.

**Comment by AnirudhDagar** [comment]
Yes, because CI is still unhappy since the notebooks are timing out, even after using LeNet.

**Comment by mseeger** [comment]
@AnirudhDagar These notebooks train LeNet on FashionMNIST. I suppose since LeNet has many convolutions, this would run faster on GPU, and then maybe the 20 min limit is not triggered?

**Comment by AnirudhDagar** [comment]
Yes, that's the very reason I assumed we needed GPUs for those two notebooks. Sorry for not already enabling gpus here, I was busy working on the JAX release. I'll take a look now.

**Comment by mseeger** [comment]
One problem is the notebook in question, hyperband-intro.md, contains two runs of HPO. What I'll do is that I split this into two notebooks.

**Comment by AnirudhDagar** [comment]
Interesting, should have caught that before, that will definitely cut down the runtime. But splitting it into two notebooks and still making the chapter as a whole continuous is very important, because the readers don't care about our CI runtime limitations. For them, the reading experience should be smooth, if it is possible to keep the flow and still split the two HPO runs, that would be awesome!

@astonzhang this will be a structural change in the chapter contents, what do you think?

**Comment by mseeger** [comment]
I think this is OK. Hyperband comes with its own explanation, own extra code, and own "experiment" run.
The only thing is that in general, Hyperband does not add much (if at all) over successive halving in practice.
But it has been marketed quite well by its authors, so we decided to include it.

**Comment by mseeger** [comment]
OK. I have doubts about the Hyperband implementation. What I'll do is that I'll remove it. We are still totally fine without it, we have successive halving, both implemented from scratch and using Syne Tune.

We can bring back Hyperband later, at this point there is no big loss.

**Comment by mseeger** [comment]
I will be out for a bit, but come back to this later in the evening. I am confident we can make this work.

@AnirudhDagar Is the CI running GPUs now?

**Comment by AnirudhDagar** [comment]
@mseeger not yet, I'll try to add that before you are back later.

**Comment by mseeger** [comment]
OK, the notebooks are running fine for me now. CI seems not operating right now.

**Comment by AnirudhDagar** [comment]
We can remove this unused variable `num_gpus=1` later when I fix the problem at [source](https://github.com/d2l-ai/d2l-book/blob/a1ce88f3894cf9b5714189fd774b5fb9d9a7c6a6/d2lbook/resource.py#L30-L46) implementing a GPU mark at the top of the notebooks next week. Right now I don't have the bandwidth to fix that in source. For now, we can keep this to fake the requirement of GPU in `sh-intro` for d2lbook.

**Comment by astonzhang** [comment]
@d2l-bot plz rebuild

**Comment by mseeger** [comment]
OK, working on this again

**Comment by mseeger** [comment]
@d2l-bot plz rebuild

**Comment by mseeger** [comment]
@AnirudhDagar Do you know why the CI does not run?

**Comment by AnirudhDagar** [comment]
Right now, CI is already running two other tasks (from jax branch i think). We limit running two jobs simultaneously since we only have 4 GPUs on our server and if more than two jobs run at a time, we may end up with out-of-memory issues.

**Comment by AnirudhDagar** [comment]
Ok, it should be evaluating the latest commit now

**Comment by mseeger** [comment]
Damn! All the notebooks ran through in less than 13 mins, but there was a mistake in d2l.bib.

**Comment by AnirudhDagar** [comment]
Haha yes, I was just looking at that. Now the notebooks won't run again, since they are in cache. So CI will be fast!!

File: chapter_hyperparameter_optimization/hyperopt-api.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (mseeger) [reply chapter_hyperparameter_optimization/hyperopt-api.md]: Done

**Comment by mseeger** [comment]
OK, I am signing off. This should go through now, and if not, nothing I can do

**Comment by astonzhang** [comment]
Thanks for all your amazing work! To save everyone's time I've added my edits directly on the PR. In the future (including the following TODO), please raise PR directly on the master branch.


TODO @mseeger (by EOD today. Time estimation: ~15 mins)

For each notebook: please create a discussion post under category "English Version - pytorch" (you'll need to sign up an account) at https://discuss.d2l.ai/c/english-version/5 and add that post at the end of the corresponding notebook

Example:

At the end of https://raw.githubusercontent.com/d2l-ai/d2l-en/master/chapter_linear-classification/softmax-regression-scratch.md

we add (plz remove the [topic name] in the URL: discuss.d2l.ai/t/[topic name]/[topic id] -> discuss.d2l.ai/t/[topic id])

```
:begin_tab:`pytorch`
[Discussions](https://discuss.d2l.ai/t/51)
:end_tab:
```
where https://discuss.d2l.ai/t/softmax-regression-implementation-from-scratch/51 was created correspondingly



TODOs @aaronkl  (next week if possible)
*  Add exercises for every notebook
*  Fig 19.3.1, 19.5.1, 19.5.2: Apply correct font size in https://github.com/d2l-ai/d2l-en/blob/master/STYLE_GUIDE.md#figure. Use transparent background. Upload graffle files to graffle/hpo/ on d2l-en repo.
* Need more space efficient display of tuner.run() output INFO:XXX in rs-async and sh-async, and  tuner.run(number_of_trials=30) output plots in sh-intro
* sh.svg cannot display properly in html so I used sh.png tentatively. We'll still need svg for this figure (need to troubleshoot).

**Comment by mseeger** [comment]
Somehow, I cannot sign up for an account for comments. Says that there were too many accesses from my IP address?

**Comment by AnirudhDagar** [comment]
Weird! Can you please try with your mobile network?

**Comment by mseeger** [comment]
OK, I have an account. I suppose I need to create a separate discussion topic for each section?
I posted one, but it needs to be reviewed. When trying to post another one, the system blocked me.

Leaving this for later.

**Comment by AnirudhDagar** [comment]
Approved!

**Comment by mseeger** [comment]
OK, I created 4 topics, but now I am blocked again. How urgent is this?

I created a new branch, which does two things:
- Reduce log output for tuner.run()
- Add these discussion things at the end (except for one notebook)

Do you want me to create a PR with this now, or can this wait until I maybe thought about some exercises?
